core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/UpxCompressionBuildStep.java

```diff
@@ -0,0 +1,204 @@
+package io.quarkus.deployment.pkg.steps;
+
+import static io.quarkus.deployment.pkg.steps.LinuxIDUtil.getLinuxID;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.commons.lang3.SystemUtils;
+import org.jboss.logging.Logger;
+
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.pkg.NativeConfig;
+import io.quarkus.deployment.pkg.builditem.ArtifactResultBuildItem;
+import io.quarkus.deployment.pkg.builditem.NativeImageBuildItem;
+import io.quarkus.deployment.util.FileUtil;
+import io.quarkus.deployment.util.ProcessUtil;
+
+public class UpxCompressionBuildStep {
+
+    private static final Logger log = Logger.getLogger(UpxCompressionBuildStep.class);
+
+    /**
+     * The name of the environment variable containing the system path.
+     */
+    private static final String PATH = "PATH";
+
+    @BuildStep(onlyIf = NativeBuild.class)
+    public void compress(NativeConfig nativeConfig, NativeImageBuildItem image,
+            BuildProducer<ArtifactResultBuildItem> artifactResultProducer) {
+        if (nativeConfig.compression.level.isEmpty()) {
+            log.debug("UPX compression disabled");
+            return;
+        }
+
+        Optional<File> upxPathFromSystem = getUpxFromSystem();
+        boolean inContainerUpx = upxPathFromSystem.isEmpty() && nativeConfig.isContainerBuild();
+        if (inContainerUpx) {
+            log.infof("Running UPX from a container using the builder image: " + nativeConfig.builderImage);
+            if (!runUpxInContainer(image, nativeConfig)) {
+                throw new IllegalStateException("Unable to compress the native executable");
+            }
+            return;
+        }
+
+        if (upxPathFromSystem.isPresent()) {
+            log.debug("Running UPX from system path");
+            if (!runUpxFromHost(upxPathFromSystem.get(), image.getPath().toFile(), nativeConfig)) {
+                throw new IllegalStateException("Unable to compress the native executable");
+            }
+            return;
+        }
+
+        log.errorf("Unable to compress the native executable, cannot find %s on the system path and the native" +
+                " compilation did not use an in-container build", getUpxExecutableName());
```

Comment:
This looks wrong, the message appears even when I use an in-container build

Reply:
Or perhaps the condition in https://github.com/quarkusio/quarkus/pull/21613/files#diff-7eda30d4763a6eef9349cb965a520deb4058083829d1e6e249e9b4cc16daa8d4R43 doesn't seem to work when building without having GRAAL_VM set

Reply:
This is no longer shown after the condition above was corrected